File: backend/app/schemas/public_form.py
from datetime import datetime

from pydantic import BaseModel, ConfigDict, Field, field_validator


class PublicFormLinkRead(BaseModel):
    form_id: str
    status: str
    public_slug: str | None
    public_url: str | None
    api_url: str | None


class PublicFormOptionRead(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: str
    label: str
    value: str
    sort_order: int


class PublicFormQuestionRead(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: str
    section_id: str | None
    instrument_id: str | None
    dimension_id: str | None
    project_variable_id: str | None
    code: str | None
    label: str
    help_text: str | None
    question_type: str
    question_role: str
    measurement_level: str
    data_type: str
    is_required: bool
    is_scored: bool
    is_reverse_scored: bool
    min_value: float | None
    max_value: float | None
    sort_order: int
    options: list[PublicFormOptionRead]


class PublicFormDimensionRead(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: str
    name: str
    code: str | None
    sort_order: int


class PublicFormInstrumentRead(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: str
    name: str
    acronym: str | None
    dimensions: list[PublicFormDimensionRead]


class PublicFormSectionRead(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: str
    title: str
    description: str | None
    sort_order: int


class PublicFormRead(BaseModel):
    id: str
    project_id: str
    title: str
    description: str | None
    instructions: str | None
    status: str
    sections: list[PublicFormSectionRead]
    instruments: list[PublicFormInstrumentRead]
    questions: list[PublicFormQuestionRead]
    thank_you_message: str | None
    metadata_json: str | None = None


class PublicAnswerCreate(BaseModel):
    question_id: str
    option_id: str | None = None
    value_text: str | None = None
    value_number: float | None = None
    value_date: datetime | None = None
    value_json: list[str] | list[dict] | dict | None = None

    @field_validator("value_text")
    @classmethod
    def normalize_optional_text(cls, value: str | None) -> str | None:
        if value is None:
            return None
        cleaned = value.strip()
        return cleaned or None


class PublicFormResponseCreate(BaseModel):
    respondent_code: str | None = Field(default=None, max_length=100)
    answers: list[PublicAnswerCreate] = Field(default_factory=list)
    metadata_json: dict | list | None = None

    @field_validator("respondent_code")
    @classmethod
    def normalize_optional_text(cls, value: str | None) -> str | None:
        if value is None:
            return None
        cleaned = value.strip()
        return cleaned or None


class PublicFormResponseRead(BaseModel):
    status: str
    response_id: str
    form_id: str
    submitted_at: datetime

File: frontend/src/types/publicForm.ts
export interface PublicFormSectionRead {
  id: string;
  title: string;
  description?: string | null;
  sort_order: number;
}

export interface PublicFormDimensionRead {
  id: string;
  name: string;
  code?: string | null;
  sort_order: number;
}

export interface PublicFormInstrumentRead {
  id: string;
  name: string;
  acronym?: string | null;
  dimensions: PublicFormDimensionRead[];
}

export interface PublicFormOptionRead {
  id: string;
  label: string;
  value: string;
  sort_order: number;
}

export interface PublicFormQuestionRead {
  id: string;
  section_id?: string | null;
  instrument_id?: string | null;
  dimension_id?: string | null;
  project_variable_id?: string | null;
  code?: string | null;
  label: string;
  help_text?: string | null;
  question_type: string;
  question_role: string;
  measurement_level: string;
  data_type: string;
  is_required: boolean;
  is_scored: boolean;
  is_reverse_scored: boolean;
  min_value?: number | null;
  max_value?: number | null;
  sort_order: number;
  options: PublicFormOptionRead[];
}

export interface PublicFormRead {
  id: string;
  project_id: string;
  title: string;
  description?: string | null;
  instructions?: string | null;
  status: string;
  sections: PublicFormSectionRead[];
  instruments: PublicFormInstrumentRead[];
  questions: PublicFormQuestionRead[];
  thank_you_message?: string | null;
  metadata_json?: string | null;
}

export interface PublicAnswerCreate {
  question_id: string;
  option_id?: string | null;
  value_text?: string | null;
  value_number?: number | null;
  value_date?: string | null;
  value_json?: any;
}

export interface PublicFormResponseCreate {
  respondent_code?: string | null;
  answers: PublicAnswerCreate[];
  metadata_json?: any;
}

export interface PublicFormResponseRead {
  status: string;
  response_id: string;
  form_id: string;
  submitted_at: string;
}
